barretenberg/cpp/src/barretenberg/circuit_checker/circuit_checker.cpp

```diff
@@ -0,0 +1,190 @@
+#include "circuit_checker.hpp"
+#include "barretenberg/flavor/goblin_ultra.hpp"
+#include <barretenberg/plonk/proof_system/constants.hpp>
+#include <unordered_set>
+
+namespace bb {
+
+template <> auto CircuitChecker::init_empty_values<UltraCircuitBuilder_<UltraArith<bb::fr>>>()
+{
+    return UltraFlavor::AllValues{};
+}
+
+template <> auto CircuitChecker::init_empty_values<GoblinUltraCircuitBuilder_<bb::fr>>()
+{
+    return GoblinUltraFlavor::AllValues{};
+}
+
+template <typename Builder> bool CircuitChecker::check(const Builder& builder_in)
+{
+    // Create a copy of the input circuit and finalize it
+    Builder builder{ builder_in };
+    builder.finalize_circuit();
+
+    // Construct a hash table for lookup table entries to efficiently determine if a lookup gate is valid
+    LookupHashTable lookup_hash_table;
+    for (const auto& table : builder.lookup_tables) {
+        const FF table_index(table.table_index);
+        for (size_t i = 0; i < table.size; ++i) {
+            lookup_hash_table.insert({ table.column_1[i], table.column_2[i], table.column_3[i], table_index });
+        }
+    }
+
+    // Instantiate structs used for checking tag and memory record correctness
+    TagCheckData tag_data;
+    MemoryCheckData memory_data{ builder };
+
+    // Initialize empty AllValues of the correct Flavor based on Builder type; for input to Relation::accumulate
+    auto values = init_empty_values<Builder>();
+    Params params;
+    params.eta = memory_data.eta; // used in Auxiliary relation for RAM/ROM consistency
+
+    // TODO(https://github.com/AztecProtocol/barretenberg/issues/867): Once we sort gates into their respective blocks
+    // we'll need to either naively run this on all blocks or run only the relevant checks on each block.
+    auto& block = builder.blocks.main;
+
+    // Perform checks on each gate defined in the builder
+    bool result = true;
+    for (size_t idx = 0; idx < block.size(); ++idx) {
+        populate_values(builder, block, values, tag_data, memory_data, idx);
+
+        result = result && check_relation<Arithmetic>(values, params);
+        result = result && check_relation<Elliptic>(values, params);
+        result = result && check_relation<Auxiliary>(values, params);
+        result = result && check_relation<GenPermSort>(values, params);
+        result = result && check_lookup(values, lookup_hash_table);
+        if constexpr (IsGoblinBuilder<Builder>) {
+            result = result && check_relation<PoseidonInternal>(values, params);
+            result = result && check_relation<PoseidonExternal>(values, params);
+        }
+    }
+
+    // Tag check is only expected to pass after all gates have been processed
+    result = result && check_tag_data(tag_data);
+
+    return result;
+};
+
+template <typename Relation> bool CircuitChecker::check_relation(auto& values, auto& params)
```

Comment:
I'd add a description and mention that this only works for linearly independent relations